Sources/RabbitMq/Connection/BasicConnection.swift

```diff
@@ -0,0 +1,100 @@
+import AMQPClient
+import AsyncAlgorithms
+import Foundation
+import Logging
+import NIO
+import NIOSSL
+import ServiceLifecycle
+
+public actor BasicConnection: Connection {
+    private var url: String
+    private var tls: TLSConfiguration?
+    private let eventLoop: EventLoop
+    public let logger: Logger  // shared to users of Connection
+
+    private var channel: AMQPChannel?
+    private var connection: AMQPConnection?
+
+    private var connecting = false
+
+    public var configuredUrl: String {
+        return url
+    }
+
+    public var isConnected: Bool {
+        if let conn = self.connection {
+            return conn.isConnected
+        }
+        return false
+    }
+
+    public init(
+        _ url: String = "",
+        tls: TLSConfiguration? = nil,
+        eventLoop: EventLoop = MultiThreadedEventLoopGroup.singleton.next(),
+        logger: Logger = Logger(label: "\(BasicConnection.self)")
+    ) throws {
+        self.url = url
+        self.tls = tls
+        self.eventLoop = eventLoop
+        self.logger = logger
+    }
+
+    // Method to use to connect without monitoring
+    public func connect() async throws {
+        if isConnected || connecting {
+            return
+        }
+
+        // Guarded by this flag on the actor
+        connecting = true
+        defer { connecting = false }
+
+        // Actually connect
+        logger.info("Connecting to broker at \(url)")
+        connection = try await AMQPConnection.connect(
+            use: eventLoop,
+            from: AMQPConnectionConfiguration(url: url, tls: tls)
+        )
+        logger.info("Connected to broker at \(url)")
+    }
+
+    public func reconfigure(with url: String, tls: TLSConfiguration? = nil) async {
+        // If the URL changes
+        if url != self.url {
+            logger.debug("Received call to reconfigure connection from \(self.url) -> \(url)")
+
+            // Close any existing connection
+            await close()
+        }
+
+        // Update configuration
+        self.url = url
+        self.tls = tls
+    }
+
+    public func getChannel() async throws -> AMQPChannel? {
+        // Not connected
+        guard isConnected else {
+            return nil
+        }
+
+        // We're connected, let's reuse the channel
+        guard let channel = self.channel, channel.isOpen else {
```

Comment:
This can race

Reply:
Maybe I can do something similar to set a flag on the actor as on the connect method. Actually now connect() isn't typically called from different places, but still good to guard both methods just in case.

Reply:
I've implemented the async semaphore for getChannel(). I'm hoping this can protect again the data races.